### PR TITLE
Allow rewardables to show its rewards in info bar

### PR DESCRIPTION
**Pull request by rilian-la-te**
Self-explanatory.

When content cannot fit into InfoBar, creates next infobar window, which will shown after timeout.

### CODE REVIEW

**Comment by IvanSavenko** [comment]
"Status bar" is text-only area at the bottom of most windows.
I think you mean "Info Bar" - bottom-right box on adventure map?

**Comment by rilian-la-te** [comment]
Yes, info bar, like in HD mod. But I want screen from Library of Enlightenment from HD mod, to look how they show more than 2 components in InfoBar.

File: client/CPlayerInterface.cpp
Comment (IvanSavenko) [inline client/CPlayerInterface.cpp:1034]: Use std::vector instead of std::pair

**Comment by krs0** [comment]
> "Status bar" is text-only area at the bottom of most windows. I think you mean "Info Bar" - bottom-right box on adventure map?

Yup, he means the "Info Window" :)

File: client/CPlayerInterface.cpp
Comment (rilian-la-te) [reply client/CPlayerInterface.cpp:1034]: @IvanSavenko why? I cannot show more than 2 objects now(

Comment (IvanSavenko) [reply client/CPlayerInterface.cpp:1034]: You can show 0, 1 or 2 components now.
Vector would be more suitable instead of two interconnected pairs.

Besides - I don't think that server logic should be aware of how many objects can be shown in UI.
If you wish - drop "assert(size <= 2)" in UI loading.
Also, if somebody were to upgrade this code to support 3+ elements then this means rewriting entire code since you can't just store third element in pair.

If standard library had "vector with specific max capacity" class I would've used it. But it does not.

Comment (rilian-la-te) [reply client/CPlayerInterface.cpp:1034]: @IvanSavenko will try to rewrite it using CComponentBox, maybe will work for 3+ components.

Comment (IvanSavenko) [reply client/CPlayerInterface.cpp:1034]: Not sure if CComponentBox would change anything.
And I think we should use vector even with your current version.

Also, if you want to fit more elements - CComponent constructor accepts "size". You can use smaller image if you have over 2 components to display

Comment (rilian-la-te) [reply client/CPlayerInterface.cpp:1034]: @IvanSavenko I do not want to write 8 constructors for each variant. I will try CComponentBox.

**Comment by IvanSavenko** [comment]
Also - apart from Library there are also Stables (0 components), and revisiting previously visited objects, including Warrior's Tomb (has components on revisiting)

**Comment by IvanSavenko** [comment]
Also, currently this setting should be configured per-object. Sure, you can made simple mod that covers H3 objects, but it won't cover objects added by another mod.

Proposition:
1) Add default value into "config/defaultMods.json"
2) Allow objects to override default behavior - if object config has "showInInfobox" property use it instead of global default

File: client/CPlayerInterface.cpp
Comment (rilian-la-te) [reply client/CPlayerInterface.cpp:1034]: @IvanSavenko stables is 0 components only if you do not have Cavaliers.

Comment (rilian-la-te) [reply client/CPlayerInterface.cpp]: Done with vector and CComponentBox. Is there any method than:
1. Gets only first row from incoming string
2. Gets required CTextBox height for showing string without scrollbar (in given width)?

**Comment by rilian-la-te** [comment]
Maybe more radical - change InfoWindow and ShowInInfobox into one CPack with two bools than force either display? If bools are not found - up to client how display InfoWindows.

**Comment by Nordsoft91** [comment]
cool, don't forget to update wiki after merge https://github.com/vcmi/vcmi/wiki/Modding-~-Objects-~-Rewardable

**Comment by rilian-la-te** [comment]
@IvanSavenko, other objects should set this to AUTO, and then it should be adventure setting "infobar if possible", and then if there is less than 9 components and text can be packed in infobar - it will show in infobar.

**Comment by IvanSavenko** [comment]
> @IvanSavenko, other objects should set this to AUTO, and then it should be adventure setting "infobar if possible", and then if there is less than 9 components and text can be packed in infobar - it will show in infobar.

Care to test every single message in the game in every possible combination and make sure that nothing is broken?

Unless we know *for sure* that these messages can be moved into infobar without issues we should use same form as before

**Comment by rilian-la-te** [comment]
> > @IvanSavenko, other objects should set this to AUTO, and then it should be adventure setting "infobar if possible", and then if there is less than 9 components and text can be packed in infobar - it will show in infobar.
> 
> Care to test every single message in the game in every possible combination and make sure that nothing is broken?
> 
> Unless we know _for sure_ that these messages can be moved into infobar without issues we should use same form as before

I need to write a checker, like "infobar fitting possible", (less than 8 components, and can fit both text and components into infobar) - then AUTO messages with "infobar if possible is ON" should be moved to infobar.

Offtopic: why scholar and witch hut is not Rewardables? What problems they have?

**Comment by IvanSavenko** [comment]
> I need to write a checker, like "infobar fitting possible", (less than 8 components, and can fit both text and components into infobar) - then AUTO messages with "infobar if possible is ON" should be moved to infobar.

Size is not the only issue. In some cases there can be multiple messages after each other. In some cases these messages normally block game till message is closed. And there may be more cases that I am not aware of.

So please, no. 

> Offtopic: why scholar and witch hut is not Rewardables? What problems they have?

1) Unlike other objects they are configurable in h3m and not completely defined in config. 
2) They need logic where multiple values share same random value, e.g. limiter "hero has skill" should have the same value as reward "hero receives skill". Not impossible to implement but not something supported by Rewardable.

**Comment by rilian-la-te** [comment]
At least, almost all object messages should be AUTO, AFAIK, like in HD mod.

**Comment by IvanSavenko** [comment]
> At least, almost all object messages should be AUTO, AFAIK, like in HD mod.

Almost all object messages are already covered by Rewardable.
Few other exceptions (e.g. Witch Hut/Scholar/etc) should be checked manually whether infobar works with them and if yes - then set to auto.

**Comment by rilian-la-te** [comment]
@IvanSavenko mostly done (except for timers, I am not sure, if we should have configurable timer for InfoBar pickups, for GCResource it will match H3, but for other - we can have configurable timer).

File: client/adventureMap/CInfoBar.h
Comment (IvanSavenko) [inline client/adventureMap/CInfoBar.h:141]: Any need to have them public?

File: client/adventureMap/CInfoBar.cpp
Comment (IvanSavenko) [inline client/adventureMap/CInfoBar.cpp:214]:
```suggestion
	assert(numComps <= 8);
	if (numComps > 8) //Bigger than 8 components - return invalid value
		return std::numeric_limits<int>::max();
```

**Comment by rilian-la-te** [comment]
@IvanSavenko Implemented long info windows, 1.5 times bigger than usual (AUTO will use long info windows, but INFO ones do not). I want final review and maybe merge into 1.2.

File: lib/mapObjects/CRewardableObject.cpp
Comment (IvanSavenko) [inline lib/mapObjects/CRewardableObject.cpp:453]: This is fix for mana well UI bug? Where it says it gives 0 mana?
Correct solution would be to extract mana amount calculations from `CRewardableObject::grantRewardAfterLevelup` method into a separate one & use it here.

**Review by IvanSavenko — COMMENTED** [review]
Also - since translation (I presume) complete, please remove "validateTranslation" from "russian" section in mods/vcmi/mod/mod.json and check that there are no reports of missing strings on startup

File: lib/NetPacks.h
Comment (rilian-la-te) [reply lib/NetPacks.h:1156]: @IvanSavenko I need to extract it from InfoWindow then, because enum class there does not work.

Comment (IvanSavenko) [reply lib/NetPacks.h:1156]: This is why
> Optionally - move this enum to NetPacksBase.h

(And in that case name it EWindowMode / EInfoWindowMode or something similar)

File: client/adventureMap/CInfoBar.cpp
Comment (rilian-la-te) [reply client/adventureMap/CInfoBar.cpp:214]: @IvanSavenko assert is not good here, because modder can set 9 components as reward, and then we just return false from infowindow, and show modal window instead.

File: client/widgets/CComponent.h
Comment (rilian-la-te) [reply client/widgets/CComponent.h:116]: @IvanSavenko I do not know how to fit delegated constructors there. I just need these 3 values to come from constructor, not from constant, delegated constructor would not do anything other than setting this 3 values. Or you mean - write one constructor and then rewrite all other constructors to use it?

File: client/adventureMap/CInfoBar.h
Comment (rilian-la-te) [reply client/adventureMap/CInfoBar.h]: made private and method rewritten

File: client/adventureMap/CAdvMapInt.cpp
Comment (rilian-la-te) [reply client/adventureMap/CAdvMapInt.cpp]: done with adding other constructor.

File: Mods/vcmi/config/vcmi/english.json
Comment (rilian-la-te) [reply Mods/vcmi/config/vcmi/english.json]: done

File: lib/NetPacks.h
Comment (rilian-la-te) [reply lib/NetPacks.h]: done

File: client/widgets/CComponent.h
Comment (IvanSavenko) [reply client/widgets/CComponent.h:116]: Basically yes - use delegated constructor to hide default parameters.
Currently you have two constructors:
(well, two pairs, but each pair is similar:
```
CComponentBox(std::vector<std::shared_ptr<CComponent>> components, Rect position); 
CComponentBox(std::vector<std::shared_ptr<CComponent>> components, Rect position, int betweenImagesMin, int betweenSubtitlesMin, int betweenRows);
```
Constructor 1 can be converted into delegated constructor and implemented as
```
CComponentBox(std::vector<std::shared_ptr<CComponent>> components, Rect position)
: CComponentBox(components, position, 20, 10, 22)
```
(same goes for second pair of constructors, with selectable components)

Comment (rilian-la-te) [reply client/widgets/CComponent.h]: done

Comment (IvanSavenko) [reply client/widgets/CComponent.h]: > But why this is better than just 2 constructors?

In this case you basically have 2 modes:
a) create box with default size values
b) create box with custom size values

There is no point (for example) in defining betweenImagesMin but keeping default value for betweenSubtitlesMin/betweenRows.

Also, having multiple default values might lead to issues later. For example if you remove betweenSubtitlesMin code that provided 4 parameters will still work, but will 4th value will now be used for betweenRows value.
And this is not some theoretical case, but something that happened in vcmi codebase before and something that happened in company I worked at before.

File: lib/mapObjects/CRewardableObject.h
Comment (IvanSavenko) [inline lib/mapObjects/CRewardableObject.h:351]: FYI: generally, for minor changes like this you should:
1) Bump serialization version in CSerializer.h (but keep minimal version intact)
2) Serialize affected values in form
```
if ( h >= <BUMPED_VERSION>)
    h & showInInfobox;
else
    h = false;
```
In this case probably not important - will merge this PR alongside with your secondary skill changes, which bump minimal save format anyway.

**Review by IvanSavenko — APPROVED** [review]
Looks OK for me, will merge in a few days

File: lib/mapObjects/CRewardableObject.h
Comment (rilian-la-te) [reply lib/mapObjects/CRewardableObject.h:351]: If I do not need to it now, so, let's resolve this conversation? 

But what structures is saved, and what not? I need to know it)

Comment (IvanSavenko) [reply lib/mapObjects/CRewardableObject.h:351]: > But what structures is saved, and what not? 

Generally - everything from library side gets saved. There are some exceptions, but those are mostly due to inability to serialized them (e.g. std::function's).

Yes, this is extensive, but this approach works fine. In long-term, dropping serialization of handlers and serializing only map itself might be a good idea, but this is definitely not something needed soon.

Comment (rilian-la-te) [reply lib/mapObjects/CRewardableObject.h]: done, serialization version bumped

File: client/adventureMap/CInfoBar.cpp
Comment (IvanSavenko) [inline client/adventureMap/CInfoBar.cpp:330]: 1) Do you need lambda here? It is called once and is not executed in async mode at some point later
2) Personal opinion, but I would rather call it "performPush", "pushImpl", "doPush". For me "actualPush" does not sounds like a good name for some reason...
3) You worry about "optimization" too much. Oh noes, you've copied vector with 5? 10? elements which took 1 mks. But the you need to render this components which can easily take 100 or even 1000 times more than this copy. There is no point in such micro-optimizations. If you want to improve performance the you need to investigate actual timings via profiler.

**Review by IvanSavenko — APPROVED** [review]
Looks OK, merging

File: lib/mapObjects/CObjectHandler.h
Comment (rilian-la-te) [reply lib/mapObjects/CObjectHandler.h:65]: But there was converted from a bunch of static functions, so, I made it static. I do not think than we should avoid static like plague) I think if we will even need instance access, then we can convert this to non-static.

File: client/adventureMap/CInfoBar.cpp
Comment (rilian-la-te) [reply client/adventureMap/CInfoBar.cpp:330]: 1. These lambda is purely for readability. Of course, it can be changed to not using lambda, but what's wrong with it.
2. Maybe. It can be called pushImpl, for sure. Can be renamed later.
3. On huge pandoras this vector can contain around 1000 elements.

File: lib/mapObjects/CObjectHandler.h
Comment (IvanSavenko) [reply lib/mapObjects/CObjectHandler.h:65]: Static does have usage. For example singleton access.

But other than that - hardly. In my opinion static is indication of poor design.
If you have static method it is likely that this method should be either part of some namespace (and not part of class) or should be part of another class.

Whether method makes any access to class members is (normally) an implementation detail and should not be exposed to user code.

> I think if we will even need instance access, then we can convert this to non-static.

Problem is, if/when we do so, it won't be as simple as removing "static" from method declaration. This also means updating every call that used access without instance. And this part might not be trivial. No need to promote poor code that will have to be rewritten later.

Comment (rilian-la-te) [reply lib/mapObjects/CObjectHandler.h:65]: For now, there is no call of openWindow with class access notation. So, changing it to instance method is zero-cost.

Comment (nullkiller) [reply lib/mapObjects/CObjectHandler.h:65]: static/non-static is more like behavior of instance or class in general. If the method describes instance behavior but still does not use instance state it is still better to leave it instance method. But if it describes something more generic, overall class behavior, some meta-information it may be static. This thing looks like a factory method so it is ok for it to be static. Otherwise we will need some window manager

Comment (IvanSavenko) [reply lib/mapObjects/CObjectHandler.h:65]: This is not factory method. 
It generates and send network pack that asks client to show window.

And internally it uses another static member (that should not be static)

File: client/adventureMap/CInfoBar.cpp
Comment (rilian-la-te) [reply client/adventureMap/CInfoBar.cpp:330]: @IvanSavenko should I change anything here?

Comment (IvanSavenko) [reply client/adventureMap/CInfoBar.cpp:330]: Up to you.
In my opinion lambda is unnecessary, but this is very minor issue, feel free to ignore.

**Pull request merged**
